remove duplicate cluster check for delta XDS

ramaraochavali opened the pull request:
**Please provide a description of this PR:**

ramaraochavali committed:
remove duplicate cluster check for delta XDS

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | fff77fa82cf03c2ff06303a1f18d0f70da4fc96f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/54515/release-notes_istio/1874764097942196224) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test release-notes

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-ambient_istio | fff77fa82cf03c2ff06303a1f18d0f70da4fc96f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/54515/integ-ambient_istio/1874764087993307136) | true | `/test integ-ambient`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

hzxuzhonghu commented on pilot/pkg/networking/core/cluster.go:
```diff
@@ -273,7 +273,11 @@ func (configgen *ConfigGeneratorImpl) buildClusters(proxy *model.Proxy, req *mod
 	}
-	resources = cb.normalizeClusters(resources)
+	// Only normalize clusters if the proxy is not using delta xDS. For Delta XDS it is not possible to
+	// identify duplicates at Istiod.
```
Can you elaborate, though delta push part  of clusters, why cannot they have duplicate names?

ramaraochavali commented:
Becuase is delta we only send the changed service related clusters. So it
is not a foolproof way to detect duplicates unlike in Sotw

On Thu, 2 Jan 2025, 17:55 hzxuzhonghu, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pilot/pkg/networking/core/cluster.go
> <https://github.com/istio/istio/pull/54515#discussion_r1900828995>:
>
> > @@ -273,7 +273,11 @@ func (configgen *ConfigGeneratorImpl) buildClusters(proxy *model.Proxy, req *mod
>  	for _, c := range clusters {
>  		resources = append(resources, &discovery.Resource{Name: c.Name, Resource: protoconv.MessageToAny(c)})
>  	}
> -	resources = cb.normalizeClusters(resources)
> +	// Only normalize clusters if the proxy is not using delta xDS. For Delta XDS it is not possible to
> +	// identify duplicates at Istiod.
>
> Can you elaborate, though delta push part of clusters, why cannot they
> have duplicate names?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/54515#pullrequestreview-2527488492>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEDINKEYNMFNJZNJYGB7HWD2IUV4LAVCNFSM6AAAAABUPWVCFGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKMRXGQ4DQNBZGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

howardjohn commented on pilot/pkg/networking/core/cluster.go:
```diff
@@ -273,7 +273,11 @@ func (configgen *ConfigGeneratorImpl) buildClusters(proxy *model.Proxy, req *mod
-	resources = cb.normalizeClusters(resources)
+	// Only normalize clusters if the proxy is not using delta xDS. For Delta XDS it is not possible to
+	// identify duplicates at Istiod.
+	if proxy.Metadata != nil && proxy.Metadata.Raw["ISTIO_DELTA_XDS"] == "false" {
```
We should find a more reliable mechanism than Metadata.Raw. Like maybe pass a bool from BuildDeltaClusters or similar

howardjohn reviewed (commented):
Funny you came across this.. in WDS we have the same bug and I was just about to start working on it.

However, I am not sure this is the right fix. I agree our `normalizeClusters` will sometimes fail to normalize. But now we will make it so it _always_ fails. This seems worse?

Won't we now be able to get NACK from envoy from duplicates? Or worse, we will flip-flop on each push?

I was thinking we may need to dedupe at a higher level than the delta works maybe

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#54517)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#54523)

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#54519)

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#54524)

howardjohn committed:
Increase timeout on benchmark test (#54527)

We keep hitting this. We should probably find a way to make the tst
shorter, but for now bump the time

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#54532)

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#54531)

howardjohn committed:
Replace Eric with keithmattix for top level ownership (#54520)

As Eric has stepped down, I nominate Keith to take this place. Keith is one of the most active maintainers and now a member of the TOC so a good fit for these top level decisions.

istio-testing committed:
Automator: update ztunnel@master in istio/istio@master (#54534)

howardjohn committed:
Use vtproto to encode gRPC messages (#54501)

* Use vtproto to encode gRPC messages

* lint

howardjohn committed:
More WDS to use vtprotobuf (#54500)

* More WDS to use vtprotobuf

* fuzz

* bump

howardjohn committed:
istioctl: remove long experimental flag --xds-via-agents (#54484)

This was added many many years ago to support some niche use cases.
Since then, I have reached out to the folks who use it (AFAIK) and they
are not using it anymore. Given the complexity here is *very*
substantial, I think this is worth cleaning up. Additionally, this
lowers our dependency on the agent.

Fixes https://github.com/istio/istio/issues/35833

hzxuzhonghu reviewed (changes requested):
IIRC this is not right, delta can fallback to full push

hzxuzhonghu commented:
FYI https://github.com/istio/istio/blob/fff77fa82cf03c2ff06303a1f18d0f70da4fc96f/pilot/pkg/networking/core/cluster.go#L74

zirain committed:
adopt envoy.extensions.wasm.v3.PluginConfig.fail_policy (#54512)

howardjohn committed:
ambient: skip alias computation when not needed (#54530)

In analysis, this is about 5% of CPU of Istiod for ambient mode (much
higher than I would expect!!). This is also before a lot of
optimizations I have open as well, so probably closer to 10% now.

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#54542)

howardjohn commented:
> > Won't we now be able to get NACK from envoy from duplicates
> 
> hmm https://github.com/envoyproxy/envoy/blob/47231ff856be0942763c40f286d3d4ee88cddda4/source/extensions/config_subscription/grpc/delta_subscription_state.cc#L196 - This behaviour is different from SoTW. In SOTW, it continues to process them only ignore the duplicated cluster.

Well we may not get NACK but we are going to flip-flop then. Each time we push we will get a random cluster. In the current SotW we deterministically pick one.

howardjohn committed:
Move resourceNames to a set (#54464)

Address comments

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#54547)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#54548)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#54552)

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#54550)

howardjohn committed:
Benchmark WDS (#54549)

istio-testing committed:
Automator: update ztunnel@master in istio/istio@master (#54554)

howardjohn committed:
protobuf: regen with new tooling (#54553)

* protobuf: regen with new tooling

* update common

ramaraochavali commented:
> I was thinking we may need to dedupe at a higher level than the delta works maybe

What are your thoughts when you say "at a higher level"? Wondering if we can use resourceNames to identify whether we have sent such a cluster earlier or not?

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#54558)

istio-testing committed:
Automator: update ztunnel@master in istio/istio@master (#54559)

ramaraochavali committed:
remove expensive sprintf in configs updated (#54555)

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

ramaraochavali committed:
fix test flake (#54556)

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

ramaraochavali committed:
unify xds push needed checks (#54451)

* unify xds push needed checks

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* review comments

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* more comments

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* named return

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* add waypoint check

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

---------

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

howardjohn committed:
Ambient: Add pre-marshalling of xds (#54521)

* ambient: pre-marshal XDS resources

* lint

* lint:

* add test

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#54560)

howardjohn commented:
I was thinking like we pre-filter out the services or something.. but it may be especially hard when we consider EnvoyFilters which can add or remove, etc. Not sure a great approach tbh.

howardjohn committed:
delta xds: fix regression in pushing on unsubscribe (#54563)

This was caught by e2e tests but not unit tests. I fixed the bug and
added proper unit tests (we previously had this test for sotw but not
delta!)

MaxBab committed:
Refactor Telemetry tests "IsKindCluster" function (#54507)

The "isKindCluster" function used to detect if the cluster during the
test is of "Kind" type or any other cluster type - vanilla Kubernetes or
Openshift.

Move the "isKindCluster" function under the
"(resource.Context).Clusters()" to be available as
"(resource.Context).Clusters().IsKindCluster()"

Signed-off-by: Maxim Babushkin <mbabushk@redhat.com>

bleggett committed:
Handle out-of-order ztunnel connection removals correctly. (#54565)

* Fix #54544 (and probably #53843)

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

* Relnote

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

* Add tests, just remove `latestConn`

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

* Naming

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

* Don't race over global config var

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

* Remove commented cruft

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

---------

Signed-off-by: Benjamin Leggett <benjamin.leggett@solo.io>

keithmattix committed:
Fix Ambient Peer Authn flake (#54566)

* Peer authn flake fix

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Add release note

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

---------

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#54577)

GabrielAlacchi committed:
Resolve JWKS Resolver TOCTOU Race Conditions (#54118)

hzxuzhonghu committed:
proxy network label should be set based on pod labels, system namespa… (#54513)

* proxy network label should be set based on pod labels, system namespace labels and meshNetwork config

* update

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#54583)

keithmattix committed:
Sort portlevel mtls before converting (#54584)

* Sort portlevel mtls before converting

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

* Create iter.Seq for stable map iteration

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

---------

Signed-off-by: Keith Mattix II <keithmattix@microsoft.com>

Winbobob committed:
Bump golang.org/x/net to v0.33.0 (#54585)

istio-testing committed:
Automator: update ztunnel@master in istio/istio@master (#54573)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#54578)

howardjohn committed:
ambient: drop dedupe of workloads (#54564)

* ambient: drop dedupe of workloads

```
                                                           │ /tmp/baseline │              /tmp/remove              │
                                                           │    sec/op     │   sec/op     vs base                  │
AddressFullGeneration/serviceentry-workloadentry-16          1223.3µ ± 60%   705.3µ ± 5%  -42.35% (p=0.000 n=10+6)
AddressIncrementalGeneration/serviceentry-workloadentry-16    1.583µ ± 34%   1.079µ ± 7%  -31.82% (p=0.000 n=10+6)
geomean                                                       44.00µ         27.59µ       -37.30%

                                                           │ /tmp/baseline │               /tmp/remove                │
                                                           │     B/op      │     B/op      vs base                    │
AddressFullGeneration/serviceentry-workloadentry-16          1006.7Ki ± 0%   869.0Ki ± 0%  -13.67% (p=0.000 n=10+6)
AddressIncrementalGeneration/serviceentry-workloadentry-16      968.0 ± 0%     968.0 ± 0%        ~ (p=1.000 n=10+6) ¹
geomean                                                       30.85Ki        28.66Ki        -7.09%

                                                           │ /tmp/baseline │              /tmp/remove               │
                                                           │   allocs/op   │  allocs/op   vs base                   │
AddressFullGeneration/serviceentry-workloadentry-16            3.672k ± 0%   3.623k ± 0%  -1.33% (n=10+6)
AddressIncrementalGeneration/serviceentry-workloadentry-16      13.00 ± 0%    13.00 ± 0%       ~ (p=1.000 n=10+6) ¹
geomean                                                         218.5         217.0       -0.67%
```

* fix network lookup

ramaraochavali commented:
I looked at it a bit more. In delta, we rely on the ServiceForHostName on sidecar and that already dedupes https://github.com/istio/istio/blob/956f765af7c1c6b8bf47ebc1d924e4d34853ebe7/pilot/pkg/model/sidecar.go#L958 - So I think we can remove the normalizeClusters? Or is there any other possibility?

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#54600)

howardjohn commented:
At the very least EnvoyFilter can mess with things. Not sure of any other issues though since it does look like we should have deduplicated from the services already

Commit:
Automator: update common-files@master in istio/istio@master (#54594)

Co-authored-by: istio-testing <istio-testing-bot@google.com>

howardjohn committed:
Implement `serviceAccounts` matcher for authz policy (#54091)

* Service Account Authz Policy implementation

* fix tests

* fixes

istio-testing committed:
Automator: update common-files@master in istio/istio@master (#54603)

istio-testing committed:
Automator: update proxy@master in istio/istio@master (#54605)

istio-testing committed:
Automator: update ztunnel@master in istio/istio@master (#54606)

ramaraochavali committed:
Merge branch 'fix/remove_duplicate_clusters' of https://github.com/ramaraochavali/istio into fix/remove_duplicate_clusters

ramaraochavali committed:
only dedupe envoy filter clusters

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

ramaraochavali commented:
Ok. Changed to dedupe only envoy filter and inbound clusters. I think this should be safe fix.

howardjohn commented:
I think you have a bad merge

ramaraochavali commented:
https://github.com/istio/istio/pull/54614 - Corrected PR

ramaraochavali commented:
Closing this as other PR is opened for it
